src/monodroid/jni/strings.hh

```diff
@@ -0,0 +1,731 @@
+#ifndef __STRINGS_HH
+#define __STRINGS_HH
+
+#include <array>
+#include <cstring>
+#include <cerrno>
+#include <limits>
+#include <type_traits>
+#include <unistd.h>
+
+#include "platform-compat.hh"
+#include "logger.hh"
+
+namespace xamarin::android::internal
+{
+	constexpr size_t SENSIBLE_TYPE_NAME_LENGTH = 128;
+	constexpr size_t SENSIBLE_PATH_MAX = 512;
+	// 64-bit unsigned or 64-bit signed with sign
+	constexpr size_t MAX_INTEGER_DIGIT_COUNT_BASE10 = 21;
+
+#if DEBUG
+	constexpr bool BoundsCheck = true;
+#else
+	constexpr bool BoundsCheck = true;
+#endif
+
+	class string_segment
+	{
+	public:
+		force_inline bool initialized () const noexcept
+		{
+			return !_fresh;
+		}
+
+		force_inline const char* start () const noexcept
+		{
+			return _start;
+		}
+
+		force_inline size_t length () const noexcept
+		{
+			return _length;
+		}
+
+		force_inline bool empty () const noexcept
+		{
+			return length () == 0;
+		}
+
+		force_inline bool equal (const char *s) const noexcept
+		{
+			if (s == nullptr)
+				return false;
+
+			return equal (s, strlen (s));
+		}
+
+		force_inline bool equal (const char *s, size_t s_length) const noexcept
+		{
+			if (!can_access (s_length)) {
+				return false;
+			}
+
+			if (_length != s_length) {
+				return false;
+			}
+
+			if (_length == 0) {
+				return true;
+			}
+
+			return memcmp (_start, s, length ()) == 0;
+		}
+
+		force_inline bool starts_with (const char *s) const noexcept
```

Comment:
I feel like there should be an overload for string constants here, too:

```cpp
template<size_t Size>
force_inline bool starts_with (const char (&s)[Size]) const noexcept
{
    return starts_with (s, Size);
}
```

This avoids the `strlen(s)` call.